### PR TITLE
Make it possible to invoke multiple tests at once

### LINKED ISSUE #31401: Test: FreeCAD is missing a mechanism for running multiple groups of tests at once
### Problem description

You can only provide `-t` once. If you want to run multiple tests, you're out of luck unless you're willing to run an entire existing test suite.

I ran into this problem during CAM development. On my branch, I wanted to routinely run `CAMTests/TestAreaOperations.py` to check for regressions and also a new file with tests for the new features in my PR. I couldn't find any way to do that without running all of CAM tests, which would have been excessively noisy and wasted a lot of time and compute. Instead, I alternated back and forth, and have now created this ticket and a PR to add support for multiple `-t` flags.

### Workbench affected?

None

### Steps to reproduce

Run FreeCAD with multiple `-t` flags, for example:

```
./bin/FreeCAD -t CAMTests.TestPathAdaptive -t CAMTests.TestAreaOperations --console
```

Alternatively mix `-t` and `-r` flags:
```
./bin/FreeCAD -t CAMTests.TestPathAdaptive -r CAMTests.TestAreaOperations --console
```

### Expected behavior

It should run all of the specified tests

### Actual behavior

It complains that that you can only use `-t` once, or if using one each of `-t` and `-r` it runs only the `-r` test.

### Development version About Info (in Safe Mode)

```shell
OS: Arch Linux (wayland)
Architecture: x86_64
Version: 26.3.0dev.45870 (Git)
Build date: 2026/06/24 11:52:30
Build type: Release
Branch: main
Hash: 67753353764d02f3cf57e9646cc3432882febc3d
Python 3.14.5, Qt 6.11.1, Coin 4.0.8, Vtk 9.6.2, boost 1_91, Eigen3 5.0.1, PySide 6.11.1
shiboken 6.11.1, SMESH 7.7.1.0, xerces-c 3.3.0, Clipper2 , OCC 7.9.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: unset/FreeCAD Classic/
Navigation Style/Orbit Style/Rotation Mode: CAD/Trackball/Drag at cursor
Logical DPI/Physical DPI/Pixel Ratio: 96/140.007/1
Installed mods:
  * SheetMetal Workbench 0.7.55
  * UserStartupScript
```

### Last known good version (optional)

```shell
none
```

### PR DESCRIPTION
**Pull request by davidgilkaufman**
Fixes: #31401

This PR makes it possible to run multiple groups of tests by specifying the `-t` or `-r` flag multiple times.

```
$ ./bin/FreeCAD -t CAMTests.TestPathAdaptive --console
...
Ran 16 tests in 3.820s

```

```
$ ./bin/FreeCAD -t CAMTests.TestAreaOperations --console
...
Ran 15 tests in 0.001s
```

```
$ ./bin/FreeCAD -t CAMTests.TestPathAdaptive -t CAMTests.TestAreaOperations --console
...
Ran 31 tests in 3.821s
```

### CODE REVIEW

**Comment by Connor9220** [comment]
This looks good.

File: src/App/Application.cpp [hunk 1]
```diff
@@ -2659,15 +2659,23 @@ void processProgramOptions(const boost::program_options::variables_map& vm, std:
...
-        else if (testCase.empty()) {
-            testCase = "TestApp.PrintAll";
+
+        if (testCases.size() == 1) {
+            if (testCases[0] == "0") {
+                testCases[0] = "TestApp.All";
+            }
+            else if (testCases[0].empty()) {
+                testCases[0] = "TestApp.PrintAll";
+            }
```
Comment (Lgt2x) [inline src/App/Application.cpp]: How do we want to handle the edge case `./bin/FreeCAD -t CAMTests.TestPathAdaptive -t 0`? Currently it only runs `CAMTests.TestPathAdaptive`, but I think it should be running all the tests.

In a similar fashion:
```
./bin/FreeCADCmd -t 0 -t 0
=> Error
```

```
./bin/FreeCADCmd -t -t 0
=> Error
```

Comment (davidgilkaufman) [reply src/App/Application.cpp]: I have no strong opinions, and I'm happy to update this PR to meet ~any simple/reasonable spec. Imo it's important that it is possible to run multiple test suits at once, but less important to handle poorly defined/malformed sets of -t arguments in any particular way.

Perhaps if any `-t` argument is `0` then other `-t` arguments are ignored and all tests run?

I'm about to go away for the weekend, but I can revisit this next week.

Comment (Lgt2x) [reply src/App/Application.cpp]: > Imo it's important that it is possible to run multiple test suits at once, but less important to handle poorly defined/malformed sets of -t arguments in any particular way.

Absolutely, I just don't want them to error out unexpectedly.

> Perhaps if any -t argument is 0 then other -t arguments are ignored and all tests run?

Sounds good. And it there is a `-t` argument without a test name or `0` after and other `-t` arguments with actual test names after, we can just ignore it. eg `./bin/FreeCADCmd -t -t 0` runs all tests, `./bin/FreeCAD -t CAMTests.TestPathAdaptive -t` runs only `TestPathAdaptive`, and `./bin/FreeCAD -t` or `./bin/FreeCAD -t -t` lists tests.

**Comment by davidgilkaufman** [comment]
I implemented a different behavior:

1. If there is any bare any `-t`, then other arguments are ignored and it prints the  list of available tests
2. Otherwise, if there is any `-t 0`, then other arguments are ignored and it runs all tests
3. Otherwise, accumulate all `-t` and `-r` values and run those tests

I chose this order of priorities mostly because of how boost's argument parser handles `initial_value` together with `compose`. While parsing, when it encounters a bare `-t`, boost overwrites any already-parsed `-t` values with the initial value (`[""]`), ignoring the instruction to compose. I originally meant to make printing options with bare `-t` the lowest priority, but boost's overwrite behavior makes this challenging. I think this alternative ordering of priorities is also reasonable, and it's easy to implement, so it's what I've implemented instead.

This PR should be ready now. All of the above examples either print the list of tests or run the obvious test suite(s).

**Comment by Lgt2x** [comment]
Tested, behavior corresponds to the description above.

**Review by Lgt2x — APPROVED** [review]
